### PR TITLE
feature/2116 setup api and cache

### PR DESCRIPTION
**Pull request by jeppekroghitk**
#### Link to ticket

https://leantime.itkdev.dk/#/tickets/showTicket/2116

#### Description

- **added styling throughout plugin**
- **rewritten timeTable.js into a class, overhauled methods and added doccomments**
- **Added seperate apihandler used by timetable.js**
- **Modified blade template to better match design**
- **Loading apihandler frontend**
- **Added apihandler symlink to install script**
- **Added additional data as well as a get hook for jquery ajax**
- **Added left join to project table in order to get project name with worklogs**

#### Screenshot of the result

If your change affects the user interface you should include a screenshot of the result with the pull request.

#### Checklist

- [ ] My code is covered by test cases.
- [ ] My code passes our test (all our tests).
- [ ] My code passes our static analysis suite.
- [ ] My code passes our continuous integration process.

If your code does not pass all the requirements on the checklist you have to add a comment explaining why this change 
should be exempt from the list.

#### Additional comments or questions

Screenshot 2024-08-22 at 16 18 48

### CODE REVIEW

File: Controllers/TimeTable.php [hunk 1]
```diff
@@ -45,29 +45,25 @@ public function post(): Response
...
                 'id' => $_POST['timesheet-id'],
             ];
-
             $this->timeTableService->updateTime($values);
         } else {
             $values = [
                 'userId' => session('userdata.id'),
                 'hours' => $_POST['timesheet-hours'],
-                'workDate' => (new Carbon($_POST['timesheet-date'], session('usersettings.timezone')))->setTimezone('UTC'),
+                'workDate' => (new Carbon($_POST['timesheet-date'], session('usersettings.timezone'))),
```
Comment (turegjorup) [inline Controllers/TimeTable.php]: Assume there is a good reason to not use UTC?

File: Controllers/TimeTable.php [hunk 2]
```diff
@@ -85,10 +81,28 @@ public function post(): Response
      */
     public function get(): Response
     {
+        if (isset($_GET['getActiveTicketIdsOfPeriod'])) {
+            $startDate = filter_input(INPUT_GET, 'start', FILTER_SANITIZE_STRING);
+            $endDate = filter_input(INPUT_GET, 'end', FILTER_SANITIZE_STRING);
+            if (!$startDate || !$endDate) {
+                die(json_encode([]));
```
Comment (turegjorup) [inline Controllers/TimeTable.php]: We usually don't die in production code 😉

File: Controllers/TimeTable.php [hunk 3]
```diff
@@ -85,10 +81,28 @@ public function post(): Response
...
+            }
+
+            $startDate = (new CarbonImmutable($startDate, session('usersettings.timezone')))->setToDbTimezone();
+            $endDate = (new CarbonImmutable($endDate, session('usersettings.timezone')))->setToDbTimezone();
+
+            $data = $this->timeTableService->getUniqueTicketIds($startDate, $endDate);
+
+            $ticketIds = $data ? array_column($data, 'ticketId') : [];
+
+            die(json_encode($ticketIds));
```
Comment (turegjorup) [inline Controllers/TimeTable.php]: Please don't die

File: Templates/timetable.blade.php [hunk 4]
```diff
@@ -2,33 +2,31 @@
...
-                <button type="button" onclick="changeWeek(-1)"><i class="fa fa-arrow-left"></i> {{ __('timeTable.button_prev_week') }}</button>
-                <button type="button" onclick="changeWeek(1)">{{ __('timeTable.button_next_week') }} <i class="fa fa-arrow-right"></i></button>
-                <button class="new-button" type="button" onclick="openEditTimeLogModal()">{{ __('timeTable.button_add_time_log') }} <i class="fa fa-plus"></i></button>
+                <button type="button" class="timetable-week-prev"><i class="fa fa-arrow-left"></i> {{ __('timeTable.button_prev_week') }}</button>
+                <button type="button" class="timetable-week-next">{{ __('timeTable.button_next_week') }} <i class="fa fa-arrow-right"></i></button>
+                <button class="timetable-new-entry" type="button">{{ __('timeTable.button_add_time_log') }} <i class="fa fa-plus"></i></button>
             </div>
         </div>
-        <div class="search-bar">
+        {{--<div class="search-bar">
```
Comment (turegjorup) [inline Templates/timetable.blade.php]: Please remove the code which was commented out

File: Templates/timetable.blade.php [hunk 5]
```diff
@@ -2,33 +2,31 @@
...
             <label class="sr-only">{{ __('timeTable.search_label') }}</label>
-            <input value="{!! $currentSearchTerm !!}" type="text" class="search-input"
-                onchange="redirectWithSearchTerm(this.value)" placeholder="{{ __('timeTable.empty_search_label') }}" />
-        </div>
-        <table class="table table-striped">
+            <input value="{!! $currentSearchTerm !!}" type="text" class="timetable-search"
+                placeholder="{{ __('timeTable.empty_search_label') }}" />
+        </div>--}}
```
Comment (turegjorup) [inline Templates/timetable.blade.php]: Please remove the code which was commented out

File: assets/timeTable.css [hunk 6]
```diff
@@ -70,13 +74,43 @@ input::-webkit-inner-spin-button {
 .table td {
   position: relative;
 }
+.time-table-container .table td:nth-child(7) {
```
Comment (turegjorup) [inline assets/timeTable.css]: `nth-child(7)` works but is fragile in regards to changes to the table. Setting a class on the weekend columns and styling on that would be better.

File: Templates/timetable.blade.php [hunk 7]
```diff
@@ -2,33 +2,31 @@
...
+        <table id="timetable" class="table">
             <thead>
                 <tr>
-                    <th scope="col">{{ __('timeTable.id_table_header') }}</th>
-                    <th scope="col">{{ __('timeTable.title_table_header') }}</th>
+                    <th style="width:60%;" scope="col">{{ __('timeTable.title_table_header') }}</th>
```
Comment (turegjorup) [inline Templates/timetable.blade.php]: Would prefer all table styles to be in the stylesheet. Looging at you `style="width:60%;`

File: assets/timeTable.css [hunk 8]
```diff
@@ -70,13 +74,43 @@ input::-webkit-inner-spin-button {
...
+.time-table-container .table .tr-total {
+  background-color: #e9e9e9;
+  font-weight: bold;
+  cursor: default;
+}
+.time-table-container .table .tr-total > td {
+  padding: 7px 15px;
+}
+.time-table-container .table .tr-total td:nth-child(7),
+.time-table-container .table .tr-total td:nth-child(8) {
```
Comment (turegjorup) [inline assets/timeTable.css]: If we donøt use `nth-child` this can probably be solved with specificity

**Review by turegjorup — APPROVED** [review]
Approved. Some cleanup recommended.

File: Controllers/TimeTable.php (hunk 1 above)
Comment (jeppekroghitk) [reply Controllers/TimeTable.php]: You were onto something here. Found an issue both here and where we get tickets.
It's important that we convert dates to UTC when saving them to the database, as well as the other way around when we get them.

**Pull request merged**